blocks/graph.py

```diff
@@ -319,3 +320,84 @@ def apply_noise(graph, variables, level, seed=None):
         replace[variable] = (variable +
                              rng.normal(variable.shape, std=level))
     return graph.replace(replace)
+
+
+def collect_parameters(graph, params):
+    """Replace parameters with a single shared variable.
+
+    This can be useful if you need to calculate the full Hessian of a
+    computational graph. It replaces parameters with slices of a single
+    large vectors like
+
+    >>> from blocks.utils import shared_floatx
+    >>> W1 = shared_floatx(numpy.random.rand(10, 10))
+    >>> W2 = shared_floatx(numpy.random.rand(10, 10))
+    >>> all_params = shared_floatx(numpy.concatenate(
+    ...     [W1.get_value().flatten(), W2.get_value().flatten()]))
+    >>> W1 = all_params[:W1.size]
+    >>> W2 = all_params[W1.size:]
+
+    Parameters
+    ----------
+    graph : :class:`ComputationGraph` instance
+        The managed Theano graph in which to collect parameters.
+    params : list of Theano shared variables
+        The parameters whose values should be collected.
+
+    Returns
+    -------
+    ComputationGraph instance
+        A new Theano graph which has all the given parameters collected
+        into a single large shared variable.
+
+    Notes
+    -----
+    Note that this replacement makes the training of the model
+    significantly slower because of the large amount of Theano's
+    ``set_subtensor`` calls needed to train the model.
+
+    Examples
+    --------
+    >>> from blocks.bricks import MLP, Sigmoid
+    >>> from blocks.bricks.cost import SquaredError
+    >>> from theano import tensor
+    >>> x = tensor.matrix()
+    >>> mlp = MLP(activations=[Sigmoid(), Sigmoid()], dims=[784, 100, 784])
+    >>> cost = SquaredError().apply(x, mlp.apply(x))
+    >>> cg = ComputationGraph(cost)
+    >>> new_cg = collect_parameters(cg, cg.shared_variables)
+
+    The new graph only has a single shared variable.
+
+    >>> new_cg.shared_variables
+    [collected_params]
+
+    The bricks' variables have been replaced with reshaped segments of this
+    single shared variable.
+
+    >>> from blocks.filter import VariableFilter
+    >>> from blocks.roles import PARAMETER
+    >>> var_filter = VariableFilter(roles=[PARAMETER])
+    >>> var_filter(new_cg.variables) # doctest: +SKIP
+    [Reshape{1}.0, Reshape{1}.0, Reshape{2}.0, Reshape{2}.0]
+
+    """
+    param_values, param_sizes, param_shapes = [], [], []
+    for param in params:
+        param_values.append(param.get_value(borrow=True))
+        param_sizes.append(param_values[-1].size)
+        param_shapes.append(param_values[-1].shape)
+
+    new_params = shared_floatx_zeros(sum(param_sizes))
+    new_params.set_value(numpy.concatenate([value.flatten()
+                                            for value in param_values]))
+    new_params.name = 'collected_params'
```

Comment:
Keep in mind that the learning rate is only part of the computation graph for the *updates*, not of the one for the cost, which is what the algorithm would be dealing with, so I don't think that will be a problem.

I agree with your argument though, but it's more complicated than that, because the alternative can be very annoying and/or dangerous as well: With an autoencoder I will only retrieve one weight matrix, which might be the behavior that I expect. However, if I added noise to weight matrices in an MLP, I will retrieve the matrices without noise, and will be left without a way of retrieving the ones with noise. That is problematic. Moreover, if I used the `collect_parameters` function, then my MLP will no longer have *any* weight matrices or biases. That is very unintuitive as well. Something as simple as getting the norm of the weight matrices is now impossible, because I have no way of retrieving them from the graph.

The advantage of separating the "shared variables" from the `PARAMETER` role, instead of making the latter a subset of the other, is that it gives us more granularity (per definition). Your autoencoder example can then be solved by taking "shared variable AND `PARAMETER`", getting the noised weight matrices of an MLP is done by doing "`PARAMETER`", while the unnoised ones can be retrieved by just taking the "shared variables" filtered by their owning bricks. Same thing for collected variables: Getting the weight matrices is done with `WEIGHTS`, finding the collected ones is done by getting the shared variables again (there will be only one).

Although more flexible, I agree that the semantics are awkward, and I am not a fan of `SUBSTITUTE` either. Ideally we want a distinction between *trainable* parameters i.e. the actual Theano shared variables with respect to which we will take the gradient, and the *semantic/bricks* parameters, which will always be the weight matrix that a linear brick uses to multiply, whether it be shared, the slice of another variable, with or without noise, etc. There is a linguistic confusion here that we can do little about, but maybe we could name the former `TRAINABLE` and the latter `PARAMETER`. In most cases, a variable (e.g. a weight matrix) will be both a `PARAMETER `and will be `TRAINABLE`, but e.g. after adding noise the original variable will only be `TRAINABLE` while the noised version will be the brick's `PARAMETER`/`WEIGHTS`.